src/storage/store/engine.rs

```diff
@@ -1474,6 +1644,55 @@ impl ShardEngine {
         )
     }
 
+    /// Verifies that a FID has an active verification for a specific address and protocol.
+    ///
+    /// This function checks if a user has previously verified ownership of an address by looking for
+    /// existing VerificationAddAddress messages in their verification store. It's used to ensure that
+    /// users can only set addresses as primary addresses if they have already proven ownership through
+    /// the verification process.
+    ///
+    /// # Note
+    /// Only checks for add messages since CRDT rules ensure that verification remove messages
+    /// automatically drop the corresponding add messages from the active state.
+    pub fn verify_fid_owns_address(
+        &self,
+        fid: u64,
+        protocol: Protocol,
+        address: &[u8],
+    ) -> Result<(), MessageValidationError> {
+        // Create a filter that only checks verification messages for the specific protocol and address
+        let target_protocol = protocol as i32;
+        let target_address = address.to_vec();
+
+        let verification_filter = |message: &proto::Message| -> bool {
+            if let Some(msg_data) = &message.data {
+                if let Some(Body::VerificationAddAddressBody(verification)) = &msg_data.body {
+                    return verification.protocol == target_protocol
+                        && verification.address == target_address;
+                }
+            }
+            false
+        };
+
+        let page_result = self.stores.verification_store.get_adds_by_fid(
```

Comment:
Use `get_verification_add` which takes in an address. It's uses an address index, it's better than paginating. And then once you get the verification you can validate the protocol matches.